File: api/src/main/java/com/delivery/api/common/exception/ApiException.java
package com.delivery.api.common.exception;

import com.delivery.api.common.error.ErrorCodeIfs;
import lombok.Getter;

@Getter
public class ApiException extends RuntimeException implements ApiExceptionIfs {

    private final ErrorCodeIfs  errorCodeIfs;

    private final String errorDescription;

    public ApiException(ErrorCodeIfs errorCodeIfs) {
        super(errorCodeIfs.getDescription()); // RuntimeException 에 에러 Description 전달
        this.errorCodeIfs = errorCodeIfs;
        this.errorDescription = errorCodeIfs.getDescription();
    }

    public ApiException(ErrorCodeIfs errorCodeIfs, String errorDescription) {
        super(errorDescription); // RuntimeException 에 에러 Description 전달
        this.errorCodeIfs = errorCodeIfs;
        this.errorDescription = errorDescription;
    }

    public ApiException(ErrorCodeIfs errorCodeIfs, Throwable tx) {
        super(tx); // RuntimeException 에 에러 Description 전달
        this.errorCodeIfs = errorCodeIfs;
        this.errorDescription = errorCodeIfs.getDescription();
    }

    public ApiException(ErrorCodeIfs errorCodeIfs, Throwable tx, String errorDescription) {
        super(tx); // RuntimeException 에 에러 Description 전달
        this.errorCodeIfs = errorCodeIfs;
        this.errorDescription = errorDescription;
    }

}

File: common/src/main/kotlin/com/delivery/common/exception/ApiException.kt
package com.delivery.common.exception

import com.delivery.common.error.ErrorCodeIfs
import jdk.jfr.Description

class ApiException : RuntimeException, ApiExceptionIfs {

    override val errorCodeIfs: ErrorCodeIfs?
    override val errorDescription: String?

    constructor(errorCodeIfs: ErrorCodeIfs) : super(errorCodeIfs.getDescription()) {
        this.errorCodeIfs = errorCodeIfs
        this.errorDescription = errorCodeIfs.getDescription()
    }

    constructor(
        errorCodeIfs: ErrorCodeIfs,
        errorCodeDescription: String
    ): super(errorCodeDescription) {
        this.errorCodeIfs = errorCodeIfs
        this.errorDescription = errorCodeDescription
    }

    constructor(
        errorCodeIfs: ErrorCodeIfs,
        throwable: Throwable
    ): super(throwable) {
        this.errorCodeIfs = errorCodeIfs
        this.errorDescription = errorCodeIfs.getDescription()
    }

    constructor(
        errorCodeIfs: ErrorCodeIfs,
        throwable: Throwable,
        errorDescription: String
    ): super(throwable) {
        this.errorCodeIfs = errorCodeIfs
        this.errorDescription = errorDescription
    }


}